File: app/src/main/java/com/qwert2603/vkmessagestat/model/DataManagerImpl.java
package com.qwert2603.vkmessagestat.model;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import com.qwert2603.vkmessagestat.Const;
import com.qwert2603.vkmessagestat.VkMessageStatApplication;
import com.qwert2603.vkmessagestat.results.IntervalType;
import com.qwert2603.vkmessagestat.vkapihelper.Progress;
import com.qwert2603.vkmessagestat.vkapihelper.VkApiHelper;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import rx.Observable;
import rx.Scheduler;

public class DataManagerImpl implements DataManager {

    @Inject
    VkApiHelper mVkApiHelper;

    @Inject
    @Named(Const.IO_THREAD)
    Scheduler mIOScheduler;

    @Inject
    @Named(Const.UI_THREAD)
    Scheduler mUIScheduler;

    public DataManagerImpl() {
        VkMessageStatApplication.getAppComponent().inject(DataManagerImpl.this);
    }

    @NonNull
    @Override
    public Pair<Observable<Results>, Observable<Progress>> getMessageStatistic(@NonNull IntervalType intervalType, int value) {
        Pair<Observable<IntegerCountMap>, Observable<Progress>> messageStatistic =
                mVkApiHelper.getMessageStatistic(intervalType, value);
        Observable<IntegerCountMap> stats = messageStatistic.first.cache();
        return new Pair<>(
                Observable.zip(stats, stats.flatMap(this::getOneResults),
                        (map, listObservable) ->
                                new Results(intervalType, value, map.getTotalSum(), map.getTotalSum(), listObservable))
                        .onBackpressureBuffer()
                        .subscribeOn(mIOScheduler)
                        .observeOn(mUIScheduler),
                messageStatistic.second
                        .onBackpressureBuffer()
                        .subscribeOn(mIOScheduler)
                        .observeOn(mUIScheduler)
        );
    }

    @Override
    public void logOut() {
        mVkApiHelper.logOut();
    }

    private Observable<List<OneResult>> getOneResults(IntegerCountMap map) {
        return mVkApiHelper.getUsersById(map.keysAsList())
                .flatMap(Observable::from)
                .map(user -> new OneResult(user, map.getPercent(user.id), map.get(user.id)))
                .toSortedList((r1, r2) -> Integer.compare(r2.getQuantity(), r1.getQuantity()));
    }

}

File: app/src/main/java/com/qwert2603/vkmessagestat/model/DataManagerImpl.kt
package com.qwert2603.vkmessagestat.model

import com.qwert2603.vkmessagestat.Const
import com.qwert2603.vkmessagestat.VkMessageStatApplication
import com.qwert2603.vkmessagestat.results.IntervalType
import com.qwert2603.vkmessagestat.vkapihelper.Progress
import com.qwert2603.vkmessagestat.vkapihelper.VkApiHelper
import rx.Observable
import rx.Scheduler
import javax.inject.Inject
import javax.inject.Named

class DataManagerImpl : DataManager {

    @Inject
    lateinit var mVkApiHelper: VkApiHelper

    @Inject
    @field:[Named(Const.IO_THREAD)]
    lateinit var mIOScheduler: Scheduler

    @Inject
    @field:[Named(Const.UI_THREAD)]
    lateinit var mUIScheduler: Scheduler

    init {
        VkMessageStatApplication.getAppComponent().inject(this@DataManagerImpl)
    }

    override fun getMessageStatistic(intervalType: IntervalType, value: Int): Pair<Observable<Results>, Observable<Progress>> {
        val messageStatistic = mVkApiHelper.getMessageStatistic(intervalType, value)
        val stats = messageStatistic.first.cache()
        return Pair(
                Observable.zip(stats, stats.flatMap { this.getOneResults(it) }
                ) { map, listObservable -> Results(intervalType, value, map.totalSum, map.totalSum, listObservable) }
                        .onBackpressureBuffer()
                        .subscribeOn(mIOScheduler)
                        .observeOn(mUIScheduler),
                messageStatistic.second
                        .onBackpressureBuffer()
                        .subscribeOn(mIOScheduler)
                        .observeOn(mUIScheduler)
        )
    }

    override fun logOut() {
        mVkApiHelper.logOut()
    }

    private fun getOneResults(map: IntegerCountMap): Observable<List<OneResult>> =
            mVkApiHelper.getResultInfos(map.keys.toList())
                    .map { resultsList ->
                        resultsList
                                .map {
                                    OneResult(
                                            resultInfo = it,
                                            percent = map.getPercent(it.id.toInt()),
                                            quantity = map.getCount(it.id.toInt())
                                    )
                                }
                                .sortedByDescending { it.quantity }
                    }

}
